bundles/org.openhab.binding.amazonechocontrol/src/main/java/org/openhab/binding/amazonechocontrol/internal/handler/AccountHandler.java

```diff
@@ -792,42 +814,62 @@ private void handlePushDeviceCommand(DopplerId dopplerId, String command, String
         }
     }
 
+    // NEED TO RUN IN A DELAYED QUEUE, ELSE TOO MANY REQUESTS
     private void handlePushActivity(@Nullable String payload) {
-        JsonCommandPayloadPushActivity pushActivity = gson.fromJson(payload, JsonCommandPayloadPushActivity.class);
-
-        Key key = pushActivity.key;
-        if (key == null) {
-            return;
-        }
-
-        Connection connection = this.connection;
-        if (connection == null || !connection.getIsLoggedIn()) {
-            return;
-        }
-        Activity[] activities = connection.getActivities(10, pushActivity.timestamp);
-        Activity currentActivity = null;
-        String search = key.registeredUserId + "#" + key.entryId;
-        for (Activity activity : activities) {
-            if (StringUtils.equals(activity.id, search)) {
-                currentActivity = activity;
-                break;
-            }
-        }
-        if (currentActivity == null) {
-            return;
+        pushActivityQueue.add(payload);
+        if (pushActivityQueueRunning.compareAndSet(false, true)) {
+            queuedPushActivity();
         }
+    }
 
+    private void queuedPushActivity() {
         @Nullable
-        SourceDeviceId @Nullable [] sourceDeviceIds = currentActivity.sourceDeviceIds;
-        if (sourceDeviceIds != null) {
-            for (SourceDeviceId sourceDeviceId : sourceDeviceIds) {
-                if (sourceDeviceId != null) {
-                    EchoHandler echoHandler = findEchoHandlerBySerialNumber(sourceDeviceId.serialNumber);
-                    if (echoHandler != null) {
-                        echoHandler.handlePushActivity(currentActivity);
+        String payload = pushActivityQueue.poll();
+        if (payload != null) {
+            try {
+                JsonCommandPayloadPushActivity pushActivity = gson.fromJson(payload,
+                        JsonCommandPayloadPushActivity.class);
+
+                Key key = pushActivity.key;
+                if (key == null) {
+                    return;
+                }
+
+                Connection connection = this.connection;
+                if (connection == null || !connection.getIsLoggedIn()) {
+                    return;
+                }
+                Activity[] activities = connection.getActivities(10, pushActivity.timestamp);
+                Activity currentActivity = null;
+                String search = key.registeredUserId + "#" + key.entryId;
+                for (Activity activity : activities) {
+                    if (StringUtils.equals(activity.id, search)) {
```

Comment:
see above, remove stringutils

Reply:
Never touched this code. Will change to a simple string equals?!?

Reply:
It seems @mgeramb is right because i fully removed the StringUtils from Connection and AccountHandler and got a lot of warnings now. For sure i check for null before the equals but it seems this doesn't matter.

Reply:
`(search == null && activity.id ==null) || (search != null && search.equals(activity.id))`

would this work? if one of `search` and `activity.id` is guaranteed to be non-null, you can omit the first check.

Reply:
For this case the fact that StringUtils returns true if both variables are null is useless because it will return in any way. So only (search != null && search.equals(activity.id)) is needed but this ends in a Travis warning like @mgeramb wrote. This kind of null check seems to not recognise there is a check for null before the equals. I will take a look how other Bindings handle this, i think its a very common case.

Reply:
What is the exact warning?

Reply:
if (activity.id != null && activity.id.equals(search)) {
...

org/openhab/binding/amazonechocontrol/internal/handler/AccountHandler.java:[846,37225] Potential null pointer access: this expression has a '@Nullable' type

Reply:
Check for null on activity first. Accessing activity.id when activity is null will fail.

Reply:
if (activity != null && activity.id != null && activity.id.equals(search)) {
                        currentActivity = activity;
                        break;
                    }

The warning still appears :(

Reply:
Is it ok if i push all the changes both of you write in the comments and try to fix the warnings after it. Because there are a lot of warning and it seems this will be a lot of work to remove all the StringUtils.

Reply:
Ignore the warnings and push the code, I‘ll check what‘s wrong then.

Reply:
Was not that easy, hopefully i have not made any misstake^^